// @web3modal/core/dist/esm/src/controllers/AccountController.js
import { subscribeKey as subKey } from 'valtio/utils';
import { proxy, ref, subscribe as sub } from 'valtio/vanilla';
import { CoreHelperUtil } from '../utils/CoreHelperUtil.js';
import { BlockchainApiController } from './BlockchainApiController.js';
import { SnackController } from './SnackController.js';
const state = proxy({
    isConnected: false,
    currentTab: 0,
    tokenBalance: []
});
export const AccountController = {
    state,
    subscribe(callback) {
        return sub(state, () => callback(state));
    },
    subscribeKey(key, callback) {
        return subKey(state, key, callback);
    },
    setIsConnected(isConnected) {
        state.isConnected = isConnected;
    },
    setCaipAddress(caipAddress) {
        state.caipAddress = caipAddress;
        state.address = caipAddress ? CoreHelperUtil.getPlainAddress(caipAddress) : undefined;
    },
    setBalance(balance, balanceSymbol) {
        state.balance = balance;
        state.balanceSymbol = balanceSymbol;
    },
    setProfileName(profileName) {
        state.profileName = profileName;
    },
    setProfileImage(profileImage) {
        state.profileImage = profileImage;
    },
    setAddressExplorerUrl(explorerUrl) {
        state.addressExplorerUrl = explorerUrl;
    },
    setSmartAccountDeployed(isDeployed) {
        state.smartAccountDeployed = isDeployed;
    },
    setCurrentTab(currentTab) {
        state.currentTab = currentTab;
    },
    setTokenBalance(tokenBalance) {
        if (tokenBalance) {
            state.tokenBalance = ref(tokenBalance);
        }
    },
    async fetchTokenBalance() {
        try {
            if (state.address) {
                const response = await BlockchainApiController.getBalance(state.address);
                this.setTokenBalance(response.balances);
            }
        }
        catch (error) {
            SnackController.showError('Failed to fetch token balance');
        }
    },
    resetAccount() {
        state.isConnected = false;
        state.caipAddress = undefined;
        state.address = undefined;
        state.balance = undefined;
        state.balanceSymbol = undefined;
        state.profileName = undefined;
        state.profileImage = undefined;
        state.addressExplorerUrl = undefined;
        state.smartAccountDeployed = undefined;
        state.currentTab = 0;
    }
};
